### ISSUE TITLE
Row level write permissions improvement ideas

### ISSUE THREAD

**Issue by noliveleger** [kobotoolbox/kpi#3204]
## Description

With this new feature, users can have different permissions for each submission. 
From the example given in the Wiki: https://github.com/kobotoolbox/kpi/wiki/Row-Level-Permission-Technical-Reference,
let's assume John get 3 submissions: 
- Submission 1&2 were submitted by John
- Submission 3 was submitted by Tino

With the first iteration of the feature, if John has any one of the write partial permissions, the bulk action checkbox, the pencil and the validation status will be enable, John will be able to do some actions with the data in the UI but these actions will be rejected by the back end.

![Screen Shot 2021-05-10 at 23 22 28](https://user-images.githubusercontent.com/24226825/117753970-6a111c00-b1e7-11eb-997c-008610b11b89.png)

It would be to enable/disable CTA per row depending on the permission on that row 

![Screen Shot 2021-05-10 at 23 39 09](https://user-images.githubusercontent.com/24226825/117754751-f708a500-b1e8-11eb-8943-3f291d8de91f.png)

(or hide not avaible CTA, question for @Ig-Rebollo)

## Ideas

The front end could fetch ids for each detected permission per page, store them and compare row ids with these arrays. 
It means, the front end could make up to 3 extra calls: 

pseudo-code could be something like this

```javascript 
 // submission for current page
let submissions = Promise('/api/v2/assets/a12134/data.json?start=0&limit=100'); 
 // permissions for current project
let permissions = Promise('/api/v2/assets/a12134/permission-assignments.json'); 
let deletePermIds = []
     validatePermIds = []
     changePermIds = []

let submission_ids = getIdsfromSubmissions(submissions);

if (this.userCanPartially('delete_submissions', asset)) {
   let filters = getFilter(asset.permissions, 'delete_submissions');
   deletePermIds = Promise('/api/v2/assets/a12134/data.json?q={'$and: {'_id': {'$in': submission_ids}, filters}});
}

if (this.userCanPartially('validate_submissions', asset)) {
   let filters = getFilter(asset.permissions, 'validate_submissions');
   validatePermIds = Promise('/api/v2/assets/a12134/data.json?q={'$and: {'_id': {'$in': submission_ids}, filters}});
}

if (this.userCanPartially('change_submissions', asset)) {
   let filters = getFilter(asset.permissions, 'change_submissions');
   changePermIds = Promise('/api/v2/assets/a12134/data.json?q={'$and: {'_id': {'$in': submission_ids}, filters}});
}
```

Front end only supports one filter at a time
Example: 
```javascript 
"filters": [
  {
    "_submitted_by": {
      "$in" : [
        "tino",
        "john"
      ]
    }
  }
]
```
Each filter is joined with an `OR` operator in the back end. So, either the front end joins the filters with `$or` while fetching ids or it sends the first index of the array. 

```javascript 
   let filters = getFilter(asset.permissions, 'change_submissions');
   // filters = '{"$or": [{"_submitted_by": {"$in" : ["tino", "john"]}}]}';
   // or 
   // filters = '{"_submitted_by": {"$in" : ["tino", "john"]}}';
```

**Comment by joshuaberetta** [comment]
@noliveleger is this different to how things have now been implemented?

**Comment by noliveleger** [comment]
@joshuaberetta, indeed. The front-end code assumes that the filters are always `{"_submitted_by": {"$in" : ["tino", "john"]}}` or `{"_submitted_by": "john"}}` and searches for a match within the result.
In reality, a filter could be anything else than `_submitted_by` if (advanced) users use the API to set their permissions.
